contracts/bridge/contracts/common/message/InboundLane.sol

```diff
@@ -0,0 +1,285 @@
+        inboundLaneNonce = InboundLaneNonce(_last_confirmed_nonce, _last_delivered_nonce);
+    }
+
+    /* Public Functions */
+
+    // Receive messages proof from bridged chain.
+    //
+    // The weight of the call assumes that the transaction always brings outbound lane
+    // state update. Because of that, the submitter (relayer) has no benefit of not including
+    // this data in the transaction, so reward confirmations lags should be minimal.
+    function receive_messages_proof(
+        OutboundLaneData memory outboundLaneData,
+        bytes32 inboundLaneDataHash,
+        bytes memory messagesProof
+    ) public {
+        verify_messages_proof(hash(outboundLaneData), inboundLaneDataHash, messagesProof);
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >= outboundLaneData.messages.length * (MAX_GAS_PER_MESSAGE + GAS_BUFFER),
+            "Lane: insufficient gas for delivery of all messages"
+        );
+        receive_state_update(outboundLaneData.latest_received_nonce);
+        receive_message(outboundLaneData.messages);
+        commit();
+    }
+
+    function relayers_size() public view returns (uint256 size) {
+        if (relayersRange.back >= relayersRange.front) {
+            size = relayersRange.back - relayersRange.front + 1;
+        }
+    }
+
+    function relayers_back() public view returns (address pre_relayer) {
+        if (relayers_size() > 0) {
+            uint256 back = relayersRange.back;
+            pre_relayer = relayers[back].relayer;
+        }
+    }
+
+	// Get lane data from the storage.
+    function data() public view returns (InboundLaneData memory lane_data) {
+        uint256 size = relayers_size();
+        if (size > 0) {
+            lane_data.relayers = new UnrewardedRelayer[](size);
+            uint256 front = relayersRange.front;
+            for (uint256 index = 0; index < size; index++) {
+                lane_data.relayers[index] = relayers[front + index];
+            }
+        }
+        lane_data.last_confirmed_nonce = inboundLaneNonce.last_confirmed_nonce;
+        lane_data.last_delivered_nonce = inboundLaneNonce.last_delivered_nonce;
+    }
+
+    // storage proof issue: must use latest commitment in lightclient, cause we rm mmr root
+    function commit() public returns (bytes32) {
+        commitment = hash(data());
+        return commitment;
+    }
+
+    /* Private Functions */
+
+    // Receive state of the corresponding outbound lane.
+    function receive_state_update(uint256 latest_received_nonce) internal returns (uint256) {
+        uint256 last_delivered_nonce = inboundLaneNonce.last_delivered_nonce;
+        uint256 last_confirmed_nonce = inboundLaneNonce.last_confirmed_nonce;
+        require(latest_received_nonce <= last_delivered_nonce, "Lane: InvalidReceivedNonce");
+        if (latest_received_nonce > last_confirmed_nonce) {
+            uint256 new_confirmed_nonce = latest_received_nonce;
```

Comment:
What is the meaning of `latest_received_nonce` from source chain's OutboundLane?

Reply:
```
        // Nonce of the latest message, received by bridged chain.
        uint256 latest_received_nonce;
```